src/dune_rules/module_trie.ml

```diff
@@ -33,22 +33,34 @@ let rec find t = function
     | Some (Leaf a) -> Option.some_if (List.is_empty ps) a
     | Some (Map t) -> find t ps)
 
-let rec gen_set t ps v =
+let rec gen_set t ps v ~on_exists =
   match ps with
   | [] -> (
     match v with
     | Leaf _ -> Code_error.raise "gen_set: no top level leaf" []
     | Map m -> m)
   | p :: ps ->
     Map.update t p ~f:(fun x ->
-        if List.is_empty ps then Some v
+        if List.is_empty ps then
+          match x with
+          | None -> Some v
+          | Some c -> on_exists c
         else
           match x with
           | None -> None
           | Some (Leaf _ as leaf) -> Some leaf
-          | Some (Map m) -> Some (Map (gen_set m ps v)))
-
-let set t k v = gen_set t k (Leaf v)
+          | Some (Map m) -> Some (Map (gen_set m ps v ~on_exists)))
+
+let gen_set (type a) (t : a t) ps v =
+  let exception Duplicate of a node in
+  match gen_set t ps v ~on_exists:(fun v -> raise_notrace (Duplicate v)) with
+  | s -> Ok s
+  | exception Duplicate d -> Error d
+
+let set t k v =
```

Comment:
optional suggestion: this function could call the original (the recursive one) `gen_set` with `~on_exists` since we're throwing the `Error` result away anyway. Something like `~on_exists:(fun _ -> Code_error.raise "unexpected duplicate" [])`

Reply:
Turns out I needed to keep the old behavior of overriding the module.